### ISSUE TITLE
Multiline

### ISSUE THREAD

**Issue by eric-schleicher** [nodebox/opentype.js#164]
This is more of a questions than an issue.

Is there a reason (techical/philospical) not to support the generation of multiline text?  

i can see how this might be considered to be too high level function for open type as it (perhap not so) subtly implies the handling of handling justifications and line heights

Generally, i would expect that in most fonts there is information (either explicitly or derivable) on how to set the a base line spacing.  I'm unfamiliar on how to determine this...  any pointers would be appreciated...

I just wasn't sure if it was a possibility, or whether i need to implement this in my own code that uses opentype, or whether it might be something we could contribute, or create as a wrapper library.

Fabric has *sort of* nailed this down, but we're no longer using it because it doesn't work for our purposes, but alas multiline text is important for our application.

-Thanks for creating this **awesome** library!

**Comment by fThues** [comment]
+1, that would be great to have.

**Comment by fdb** [comment]
Multi-line text is certainly a good idea! This would probably only be supported for western languages, though (I have no experience with the layout of complex scripts, where we get into Harfbuzz territory).

**Comment by eric-schleicher** [comment]
I have some code that I could share here.  Its essentially a (very basic)
multiline object and provides control over basic justification.

Im not particularly proud of it, but a better developer might find it
useful.

Let me know

**Comment by fdb** [comment]
Not claiming I'm a better developer, but I would love to take a look :-)

**Comment by mattdesl** [comment]
I just released [opentype-layout](https://github.com/Jam3/opentype-layout) which should handle basic word wrapping and new line characters. If nothing else, the source can be used as a reference for future work. :smile:

**Comment by fdb** [comment]
Thanks @mattdesl, looks good!

I'm still pondering if this is something that falls in the scope of OpenType.js, or is something that can be handled through libraries.

**Comment by mattdesl** [comment]
I personally don't think it should be in scope of Opentype.js, as it has the potential to bloat the library, become tied to application-specific rendering (i.e. making `lineHeight` act like CSS spec rather than typical print handling) and also is very hard to properly implement (i.e. Harfbuzz).

I think it's better to push things into user-land modules since npm has become pretty widely adopted by JS devs.

:smile:

**Comment by fdb** [comment]
Well now I have a module to refer to 😄

**Comment by Harbs** [comment]
I agree that multi-line is important, and I also agree that it should probably not be part of the core. (a complete solution is a very complex problem)

@fdb I do have experience with complex scripts and I'm currently the one maintaining Apache Flex TLF.

I'm first dipping my toes into opentype.js, and I'm impressed with what you've done. I'm hoping to build a wrapper around it to emulate the FTE and hopefully allow TLF to use opentype.js to render the text at the core. I'd love to discuss my thoughts if you want to have a discussion off-line.

**Comment by kuzvac** [comment]
Any news about this feature? :)

**Comment by fdb** [comment]
Hi @kuzvac, we're not going to make this part of opentype.js: look at [opentype-layout](https://github.com/Jam3/opentype-layout) for a module that does this.

**Comment by kuzvac** [comment]
@fdb ok, thank you.
